packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,220 @@
+//  Copyright © 2021 650 Industries. All rights reserved.
+
+#import <EXUpdates/EXUpdatesAppLauncherWithDatabase.h>
+#import <EXUpdates/EXUpdatesErrorRecovery.h>
+#import <React/RCTAssert.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, EXUpdatesErrorRecoveryTask) {
+  EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate,
+  EXUpdatesErrorRecoveryTaskLaunchNew,
+  EXUpdatesErrorRecoveryTaskLaunchCached,
+  EXUpdatesErrorRecoveryTaskLaunchEmbedded,
+  EXUpdatesErrorRecoveryTaskCrash
+};
+
+static NSInteger const EXUpdatesErrorRecoveryRemoteLoadTimeoutMs = 5000;
+
+@interface EXUpdatesErrorRecovery ()
+
+@property (nonatomic, strong) NSMutableArray<NSNumber *> *pipeline;
+@property (nonatomic, assign) BOOL isRunning;
+@property (nonatomic, assign) BOOL isWaitingForRemoteUpdate;
+@property (nonatomic, assign) NSInteger remoteLoadTimeout;
+
+@property (nonatomic, strong) dispatch_queue_t errorRecoveryQueue;
+
+@property (nonatomic, strong) NSMutableArray *encounteredErrors;
+
+@end
+
+@implementation EXUpdatesErrorRecovery
+
+- (instancetype)init
+{
+  return [self initWithErrorRecoveryQueue:dispatch_queue_create("expo.controller.errorRecoveryQueue", DISPATCH_QUEUE_SERIAL)
+                        remoteLoadTimeout:EXUpdatesErrorRecoveryRemoteLoadTimeoutMs];
```

Comment:
We want this timeout to have a min value that would allow users to reasonably download an update within the time frame, since the app might crash after this timer runs out. I could see an argument for using something like `max(EXUpdatesLaunchWaitMs, EXUpdatesErrorRecoveryRemoteLoadTimeoutMs)` but I think that might be conflating two separate concepts and if we want this to be configurable we should just add a separate config value for it. I'm open to that idea but also think we could wait and see if it's wanted.